random/imx6ull-rngb.c

```diff
@@ -0,0 +1,305 @@
+	rng_esr,
+	rng_out };
+
+
+#define RNGB_START_ADDRESS 0x02284000
+#define RNGB_IRQ           (6 + 32)
+
+
+#define SELF_TEST_DONE 1U
+#define SEED_DONE      2U
+#define ERROR_OCCURRED 4U
+#define MMAP_SIZE      4096
+
+
+/* intr_st indicates status of interrupts */
+static struct {
+	volatile uint32_t *base;
+	volatile uint32_t intr_st;
+	uint32_t port;
+
+	handle_t lock;
+	handle_t cond;
+} common_rngb;
+
+
+static int rngb_intr(unsigned int n, void *arg)
+{
+	uint32_t status = *(common_rngb.base + rng_sr);
+
+	/* Self tests are done */
+	if ((status & 0x10) != 0) {
+		common_rngb.intr_st |= SELF_TEST_DONE;
+	}
+
+	/* Seeding is done*/
+	if ((status & 0x20) != 0) {
+		common_rngb.intr_st |= SEED_DONE;
+	}
+
+	/* An error occurred*/
+	if ((status & (1 << 16)) != 0) {
+		common_rngb.intr_st |= ERROR_OCCURRED;
+	}
+
+	/* Clear interrupt */
+	*(common_rngb.base + rng_cmd) = 0x20;
+	return 1;
+}
+
+
+static void systemCleanup(int level)
+{
+	if (level > 3) {
+		resourceDestroy(common_rngb.cond);
+	}
+
+	if (level > 2) {
+		resourceDestroy(common_rngb.lock);
+	}
+
+	if (level > 1) {
+		portDestroy(common_rngb.port);
+	}
+
+	if (level > 0) {
+		munmap((void *)common_rngb.base, MMAP_SIZE);
+	}
+}
+
+
+static int systemInit(void)
+{
+	const char devpath[] = "random";
+	int err;
+	oid_t dev;
+
+	common_rngb.base = mmap(NULL, MMAP_SIZE, PROT_READ | PROT_WRITE, MAP_DEVICE, OID_PHYSMEM, RNGB_START_ADDRESS);
+	if (common_rngb.base == MAP_FAILED) {
+		printf("rngb: could not map addr\n");
+		return -1;
+	}
+
+	err = portCreate(&common_rngb.port);
+	if (err != EOK) {
+		printf("rngb: could not create port\n");
+		systemCleanup(1);
+		return -1;
+	}
+
+	dev.port = common_rngb.port;
+	dev.id = 0;
+
+	err = mutexCreate(&common_rngb.lock);
+	if (err != EOK) {
+		printf("rngb: could not create mutex for rngb\n");
+		systemCleanup(2);
+		return -1;
+	}
+
+	err = condCreate(&common_rngb.cond);
+	if (err != EOK) {
+		printf("rngb: could not create cond for rngb\n");
+		systemCleanup(3);
+		return -1;
+	}
+
+	err = create_dev(&dev, devpath);
+	if (err != EOK) {
+		printf("rngb: could not create port file %s (err %d)\n", devpath, err);
+		systemCleanup(4);
+		return -1;
+	}
+
+	return 0;
+}
+
+
+static int hardwareInit(void)
+{
+	int err;
+
+	/* Reset and wait until rngb is sleeping */
+	*(common_rngb.base + rng_cmd) = 0x40;
+	do {
+		err = *(common_rngb.base + rng_sr) & 0x4;
+	} while (err == 0);
+
+	interrupt(RNGB_IRQ, rngb_intr, NULL, common_rngb.cond, NULL);
+
+	/* Run self test */
+	*(common_rngb.base + rng_cmd) = 0x1;
+	mutexLock(common_rngb.lock);
+	while (common_rngb.intr_st == 0) {
+		condWait(common_rngb.cond, common_rngb.lock, 3000);
+	}
+	mutexUnlock(common_rngb.lock);
+
+	if ((common_rngb.intr_st & ERROR_OCCURRED) != 0) {
+		printf("rngb: self-test failed\n");
+
+		systemCleanup(4);
+		return -1;
+	}
+
+	common_rngb.intr_st = 0;
+
+	/* Run seeding */
+	*(common_rngb.base + rng_cmd) = 0x2;
+	mutexLock(common_rngb.lock);
+	while (common_rngb.intr_st == 0) {
```

Comment:
Here we are waiting for whether the seeding has finished or failed and not for all statuses, so shouldn't it be?
```suggestion
	while ((common_rngb.intr_st & (SEED_DONE | ERROR_OCCURRED)) == 0) {

```